File: src/org/elixir_lang/reference/module/UnaliasedName.java
package org.elixir_lang.reference.module;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import org.elixir_lang.psi.*;
import org.elixir_lang.psi.call.Call;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static org.elixir_lang.psi.call.name.Function.ALIAS;
import static org.elixir_lang.psi.call.name.Module.KERNEL;
import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.finalArguments;
import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.hasKeywordKey;

public class UnaliasedName {
    /*
     * Public Static Methods
     */

    @Nullable
    public static String unaliasedName(@NotNull PsiNamedElement namedElement) {
        String unaliasedName;

        if (namedElement instanceof QualifiableAlias) {
            unaliasedName = unaliasedName((QualifiableAlias) namedElement);
        } else {
            unaliasedName = namedElement.getName();
        }

        return unaliasedName;
    }

    /*
     * Private Static Methods
     */

    @Nullable
    private static String down(@NotNull PsiElement element) {
        String unaliasedName = null;

        if (element instanceof QualifiableAlias) {
            PsiNamedElement namedElement = (PsiNamedElement) element;

            unaliasedName = namedElement.getName();
        }

        return unaliasedName;
    }

    @Nullable
    private static String unaliasedName(@NotNull QualifiableAlias qualifiableAlias) {
        return up(qualifiableAlias.getParent(), qualifiableAlias);
    }

    @Nullable
    private static String up(@NotNull Call call, @NotNull QualifiableAlias entrance) {
        String unaliasedName = null;

        if (call.isCalling(KERNEL, ALIAS)) {
            PsiElement[] finalArguments = finalArguments(call);

            if (finalArguments != null && finalArguments.length > 0) {
                PsiElement firstArgument = finalArguments[0];

                unaliasedName = down(firstArgument);
            }
        }

        return unaliasedName;
    }

    @Nullable
    private static String up(@Nullable PsiElement element, @NotNull QualifiableAlias entrance) {
        String unaliasedName = null;

        if (element instanceof Call) {
            unaliasedName = up((Call) element, entrance);
        } else if (element instanceof ElixirAccessExpression ||
                element instanceof QuotableArguments ||
                element instanceof QuotableKeywordList) {
            unaliasedName = up(element.getParent(), entrance);
        } else if (element instanceof ElixirMultipleAliases) {
            unaliasedName = down(entrance);
        } else if (element instanceof QuotableKeywordPair) {
            unaliasedName =up((QuotableKeywordPair) element, entrance);
        }

        return unaliasedName;
    }

    @Nullable
    private static String up(@NotNull QuotableKeywordPair element, @NotNull QualifiableAlias entrance) {
        String unaliasedName = null;

        if (hasKeywordKey(element, "as")) {
            unaliasedName = up(element.getParent(), entrance);
        }

        return  unaliasedName;
    }

}

File: src/org/elixir_lang/reference/module/UnaliasedName.kt
package org.elixir_lang.reference.module

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import org.elixir_lang.errorreport.Logger
import org.elixir_lang.psi.*
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.call.name.Function
import org.elixir_lang.psi.call.name.Function.ALIAS
import org.elixir_lang.psi.call.name.Module.KERNEL
import org.elixir_lang.psi.impl.call.finalArguments
import org.elixir_lang.psi.impl.call.maybeModularNameToModulars
import org.elixir_lang.psi.impl.hasKeywordKey

object UnaliasedName {
    fun unaliasedName(namedElement: PsiNamedElement): String? =
            if (namedElement is QualifiableAlias) {
                unaliasedName(namedElement)
            } else if (namedElement is Call && namedElement.isCalling(KERNEL, Function.__MODULE__, 0)) {
                __MODULE__
                        .reference(namedElement, useCall = null)
                        .resolve()
                        ?.let { it as? PsiNamedElement }
                        ?.name
            } else {
                namedElement.name
            }

    private tailrec fun down(element: PsiElement): String? =
            when (element) {
                is Call ->
                    element
                            .maybeModularNameToModulars(null)
                            .mapNotNull { it.name }
                            .toSet()
                            .singleOrNull()
                is ElixirAccessExpression -> {
                    val children = element.children

                    assert(children.size >= 1)

                    down(children[0])
                }
                is QualifiableAlias -> element.name
                else -> {
                    Logger.error(
                            javaClass,
                            "Don't know how to search down below ${element.javaClass} for unaliased name",
                            element
                    )

                    "?"
                }
            }

    private fun unaliasedName(qualifiableAlias: QualifiableAlias): String? =
            up(qualifiableAlias.parent, qualifiableAlias)

    private fun up(call: Call, entrance: QualifiableAlias): String? =
            if (call.isCalling(KERNEL, ALIAS)) {
                val finalArguments = call.finalArguments()

                if (finalArguments != null && finalArguments.isNotEmpty()) {
                    val firstArgument = finalArguments[0]

                    down(firstArgument)
                } else {
                    null
                }
            } else {
                entrance.name
            }

    private tailrec fun up(element: PsiElement?, entrance: QualifiableAlias): String? =
            when (element) {
                is Call ->
                    up(element, entrance)
                is QualifiedMultipleAliases ->
                    up(element, entrance)
                is ElixirAccessExpression,
                is ElixirNoParenthesesOneArgument,
                is QuotableArguments,
                is QuotableKeywordList ->
                    up(element.parent, entrance)
                is ElixirMultipleAliases ->
                    entrance.fullyQualifiedName()
                is QuotableKeywordPair ->
                    up(element, entrance)
                else ->
                    null
            }

    private fun up(element: QuotableKeywordPair, entrance: QualifiableAlias): String? =
            if (element.hasKeywordKey("as")) {
                up(element.parent, entrance)
            } else {
                null
            }
}
